### PR TITLE
Update And Fixes

### PR DESCRIPTION
**Pull request by ThisTestUser**
Same as #149, but updated.
-Fix the docs and readme
-Add back UnconditionalSwitchRemover for Zelix
-Improved ConstantFolder to remove impossible switches and aload pop combinations
-Fix HideAccess
-Remove VariableNormalizer because LocalVariableRemover is a better filter for illegal variables
-Add DuplicateRenamer (renames duplicate method/class/field names)
-Add more methods to JVMMethodProvider

### CODE REVIEW

**Comment by samczsun** [comment]
I'll take a look at this in a few days - some stuff just came up and I don't have time to commit (ha ha) to any updates right now

File: src/main/java/com/javadeobfuscator/deobfuscator/transformers/general/peephole/UnconditionalSwitchRemover.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
+/*
```
Comment (ThisTestUser) [inline src/main/java/com/javadeobfuscator/deobfuscator/transformers/general/peephole/UnconditionalSwitchRemover.java]: So this normalizer fixes Zelix string decryption, by changing 
new int[]{1,2,3}
to
int[0] = 1;
int[1] = 2;
Should we include as part of Peephole optimizer?

File: src/main/java/com/javadeobfuscator/deobfuscator/transformers/stringer/StringEncryptionTransformer.java [hunk 2]
```diff
@@ -249,7 +249,8 @@ public boolean canCheckEquality(JavaValue first, JavaValue second, Context conte
                                     MethodNode decrypterNode = innerClassNode.methods.stream().filter(mn -> mn.name.equals(m.name) && mn.desc.equals(m.desc) && Modifier.isStatic(mn.access)).findFirst().orElse(null);
                                     if (decrypterNode != null) {
                                         Context context = new Context(provider);
-                                        context.dictionary = classpath;
+                                        context.dictionary = classpath;context.constantPools = getDeobfuscator().getConstantPools();
```
Comment (Janmm14) [inline src/main/java/com/javadeobfuscator/deobfuscator/transformers/stringer/StringEncryptionTransformer.java]: duplicate code?

File: src/main/java/com/javadeobfuscator/deobfuscator/transformers/general/peephole/ConstantFolder.java [hunk 3]
```diff
@@ -231,6 +235,14 @@ public boolean transform() throws Throwable {
                                             results.add(((Number) ((LdcFrame) frame.getComparators().get(0)).getConstant()).intValue() != ((Number) ((LdcFrame) frame.getComparators().get(1)).getConstant()).intValue());
```
Comment (Janmm14) [inline src/main/java/com/javadeobfuscator/deobfuscator/transformers/general/peephole/ConstantFolder.java]: can't one define both int values used in all the ifs before all the if-else stuff? Would result in much cleaner code.

File: src/main/java/com/javadeobfuscator/deobfuscator/executor/defined/types/JavaConstructor.java [hunk 4]
```diff
@@ -60,10 +60,6 @@ else if(arg instanceof Double)
                 javaArgs.add(0, new JavaDouble((Double)arg));
             else if(arg instanceof Long)
                 javaArgs.add(0, new JavaLong((Long)arg));
-            else if(arg instanceof String)
```
Comment (samczsun) [inline src/main/java/com/javadeobfuscator/deobfuscator/executor/defined/types/JavaConstructor.java]: Why remove these?

File: src/main/java/com/javadeobfuscator/deobfuscator/transformers/general/peephole/UnconditionalSwitchRemover.java (hunk 1 above)
Comment (samczsun) [reply src/main/java/com/javadeobfuscator/deobfuscator/transformers/general/peephole/UnconditionalSwitchRemover.java]: If the side effects aren't important to decompilation and understanding how the code works, then leave it. If the side effects are misleading (i.e. makes parts of code appear to be run when they're not), then we should fix that before merging

File: src/main/java/com/javadeobfuscator/deobfuscator/transformers/normalizer/DuplicateRenamer.java [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+import com.javadeobfuscator.deobfuscator.config.TransformerConfig;
+import com.javadeobfuscator.deobfuscator.executor.Context;
+import com.javadeobfuscator.deobfuscator.executor.defined.types.JavaClass;
+import com.javadeobfuscator.deobfuscator.executor.defined.types.JavaMethod;
+import com.javadeobfuscator.deobfuscator.executor.providers.DelegatingProvider;
+import com.javadeobfuscator.deobfuscator.utils.ClassTree;
+
+@TransformerConfig.ConfigOptions(configClass = DuplicateRenamer.Config.class)
+public class DuplicateRenamer extends AbstractNormalizer<DuplicateRenamer.Config>
```
Comment (samczsun) [inline src/main/java/com/javadeobfuscator/deobfuscator/transformers/normalizer/DuplicateRenamer.java]: Why is this needed? Won't the regular Class/Field/Method normalizers achieve the same result of distinguishing members?

File: src/main/java/com/javadeobfuscator/deobfuscator/transformers/normalizer/VariableNormalizer.java [hunk 6]
```diff
@@ -1,40 +0,0 @@
...
- */
-
-package com.javadeobfuscator.deobfuscator.transformers.normalizer;
-
-import com.javadeobfuscator.deobfuscator.config.TransformerConfig;
-import com.javadeobfuscator.deobfuscator.transformers.Transformer;
-
-import java.util.concurrent.atomic.AtomicInteger;
-
-public class VariableNormalizer extends Transformer<TransformerConfig> {
```
Comment (samczsun) [inline src/main/java/com/javadeobfuscator/deobfuscator/transformers/normalizer/VariableNormalizer.java]: Why is this removed?

File: src/main/java/com/javadeobfuscator/deobfuscator/executor/defined/types/JavaConstructor.java (hunk 4 above)
Comment (ThisTestUser) [reply src/main/java/com/javadeobfuscator/deobfuscator/executor/defined/types/JavaConstructor.java]: They're not useful in any way.

File: src/main/java/com/javadeobfuscator/deobfuscator/transformers/normalizer/DuplicateRenamer.java (hunk 5 above)
Comment (ThisTestUser) [reply src/main/java/com/javadeobfuscator/deobfuscator/transformers/normalizer/DuplicateRenamer.java]: This is an alternative (for me at least) to preserve the original class/method/field names as much as possible while still making it possible to extract the files.
With the introduction of the config, allowing certain files to be excluded may solve this.
For the unconditional switch, its only used to fix ZKM's string encryption at <clinit>, as it sometimes can't be decompiled without that. A suggestion is to include the unconditionalswitch as part of ZKM string decryptor.

File: src/main/java/com/javadeobfuscator/deobfuscator/transformers/normalizer/VariableNormalizer.java (hunk 6 above)
Comment (ThisTestUser) [reply src/main/java/com/javadeobfuscator/deobfuscator/transformers/normalizer/VariableNormalizer.java]: This was removed because it was a bad solution to the illegal local variables problem. Removing them entirely worked better for decompilers.
